### PR TITLE
Fixes #23690 - update react to 16.3

### PR DESCRIPTION
**Pull request by amirfefer**
React 16.3 updates its lifecycle methods, and deprecated 
- componentWillMount
- componentWillReceiveProps
- componentWillUpdate

Katello [uses](https://github.com/Katello/katello/blob/e8ff2f4ee53ac8bae98cb3c47c3d798eca90fd03/webpack/scenes/Subscriptions/SubscriptionsTable.js#L66)  the new `getDerivedStateFromProps` lifecycle method.
Also, `react-patternfly` 2.5.1 uses react 16.3.1 as a peer dependency

### CODE REVIEW

File: webpack/assets/javascripts/react_app/components/BreadcrumbBar/components/BreadcrumbSwitcher.js [hunk 1]
```diff
@@ -9,10 +9,10 @@ import BreadcrumbSwitcherPopover from './BreadcrumbSwitcherPopover';
 import BreadcrumbSwitcherToggler from './BreadcrumbSwitcherToggler';
 
 class BreadcrumbSwitcher extends React.Component {
-  componentWillReceiveProps(nextProps) {
```
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/components/BreadcrumbBar/components/BreadcrumbSwitcher.js:12]: componentWillReceiveProps is depracted, according to react documentations:

> ...This (componentDidUpdate) is also a good place to do network requests as long as you compare the current props to previous props

**Comment by waldenraines** [comment]
I believe we need this for 1.18 and will also need the relevant packages built.

**Comment by mmoll** [comment]
> I believe we need this for 1.18

Why?

**Comment by waldenraines** [comment]
> Why?

Because it fixes an issue in katello which is dependent on the version defined here.  @amirfefer please correct me if I'm wrong.

**Comment by amirfefer** [comment]
@waldenraines 
You are correct, as long as Katello will use the new subscriptions page in the upcoming version.
so in order that we won't have two distinct react versions, I think we need it on 1.18

**Review by boaz0 — APPROVED** [review]
LGTM :+1:

**Comment by ekohl** [comment]
We have multiple libraries that are used which have a peerDependency that isn't satisfied or worse: an actual dependency on react that wouldn't be satisfied. It'd be nice if we also looked into that and provided upstream patches where needed.

**Comment by amirfefer** [comment]
[test foreman]

**Comment by waldenraines** [comment]
[test katello]

**Review by ohadlevy — APPROVED** [review]

**Comment by ohadlevy** [comment]
@ekohl do you have that list handy? I tried running it myself:
```
 npm i                                                                                                                                                git:(5651|… 

> TheForemanDevDeps@1.19.0 postinstall /home/ohad/git/foreman
> node ./script/npm_install_plugins.js

npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@1.2.0 (node_modules/fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.2.0: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})

added 84 packages from 63 contributors in 14.007s
```
and as you can see I didn't get any warning? IMHO regardless, we should merge this.
since this is required for backporting I am ok with merging as is, but imho we should merge to latest version of react which is already at 16.4.

**Comment by ekohl** [comment]
> @ekohl do you have that list handy?

Not at the moment. I'll have a look later.

I should note that I had some issues with packaging this. npm2rpm didn't understand the dependencies of d3 and I haven't had time to look into why. For some reason it can't figure out what the correct URL to download the tarballs from. There's probably some library that could do a better job than our manual way.

**Comment by ohadlevy** [comment]
@ekohl I'm not sure I understand exactly the issue you are facing? I can't imagine that react y upgrade would relate to d3?

**Comment by ekohl** [comment]
@ohadlevy it looks like I wasn't quite awake and mixing up react and patternfly. Apologies.

**Comment by ohadlevy** [comment]
@ekohl mind approving then? :)

**Comment by waldenraines** [comment]
> @ekohl mind approving then? :)

Or @tstrachota? Would like to get this merged soon so I can merge the associated katello PR https://github.com/Katello/katello/pull/7429.

File: package.json [hunk 2]
```diff
@@ -75,10 +75,10 @@
     "patternfly": "^3.42.0",
     "patternfly-react": "^2.3.3",
     "prop-types": "^15.6.0",
-    "react": "^16.2.0",
+    "react": "^16.3.1",
```
Comment (ekohl) [inline package.json:78]: This also allows react `16.4.0` which has been released. Should I package that or did you intend `~16.3.1`?

Comment (amirfefer) [reply package.json:78]: I intended 16.3.1, but I don't mind to package 16.4.0

Comment (ekohl) [reply package.json:78]: If it stays `^16.3.1` then I'll package 16.4.0 since that's what Travis will also install while testing.

**Review by ekohl — APPROVED** [review]
https://github.com/theforeman/foreman-packaging/pull/2649 is for react. react-dom is already 16.4.0.

**Comment by tbrisker** [comment]
if we need this in 1.18-stable please open a CP PR

**Comment by waldenraines** [comment]
> if we need this in 1.18-stable please open a CP PR

@amirfefer do you mind handling this?

**Pull request merged**